src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,853 @@
+{
+        int r;
+
+        /* Set F_EXCL so table add fails if the table already exists. */
+        r = sd_nfnl_nft_message_new_table(nfnl, &ruleset[++msgcnt], nft_systemd_table_name, NLM_F_EXCL | NLM_F_ACK);
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+
+        r = sd_nfnl_nft_message_new_basechain(nfnl, &ruleset[++msgcnt], nft_systemd_table_name,
+                                              "prerouting", "nat",
+                                              NF_INET_PRE_ROUTING, NF_IP_PRI_NAT_DST + 1);
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_nft_message_new_basechain(nfnl, &ruleset[++msgcnt], nft_systemd_table_name,
+                                              "output", "nat",
+                                              NF_INET_LOCAL_OUT, NF_IP_PRI_NAT_DST + 1);
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_nft_message_new_basechain(nfnl, &ruleset[++msgcnt], nft_systemd_table_name,
+                                              "postrouting", "nat",
+                                              NF_INET_POST_ROUTING, NF_IP_PRI_NAT_SRC + 1);
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        /* set to store ip address ranges we should masquerade for */
+        r = nft_new_set(&ruleset[++msgcnt], "masq_saddr", NFT_SET_INTERVAL, TYPE_IPADDR, sizeof(uint32_t));
+        if (r < 0)
+                return r;
+
+        /*
+         * map to store ip address:port pair to dnat to.  elements in concatenation
+         * are rounded up to 4 bytes.
+         *
+         * Example: ip protocol . tcp daddr is sizeof(uint32_t) + sizeof(uint32_t), not
+         * sizeof(uint8_t) + sizeof(uint16_t).
+         */
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = nft_new_map(&ruleset[++msgcnt], nft_systemd_dnat_map_name, 0,
+                        concat_types2(TYPE_INET_PROTOCOL, TYPE_INET_SERVICE), sizeof(uint32_t) * 2,
+                        concat_types2(TYPE_IPADDR, TYPE_INET_SERVICE), sizeof(uint32_t) * 2);
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_message_new_dnat_rule_pre(&ruleset[++msgcnt], "prerouting");
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_message_new_dnat_rule_out(&ruleset[++msgcnt], "output");
+        if (r < 0)
+                return r;
+
+        assert_return(msgcnt < NFT_INIT_MSGS, -EINVAL);
+        r = sd_nfnl_message_new_masq_rule(&ruleset[++msgcnt], "postrouting");
+        if (r < 0)
+                return r;
+
+        return msgcnt;
+}
+
+int fw_nftables_init(void)
+{
+        _cleanup_(sd_netlink_unrefp) sd_netlink *nfnlp;
```

Comment:
will result in memory corruption if we exit early from this function. needs to be initialized to NULL so that the automatic destructor becomes a NOP when notthing is allocated to it yet